webui/src/pages/repositories/repository/objects.jsx

```diff
@@ -107,52 +101,19 @@ const ImportButton = ({ config, repo, reference, path, onDone, onClick, variant
             ...initialState,
             inProgress: true
         })
-
+        const updateStateFromImport = ({inProgress, done, numObj}) => {
+            setImportState({inProgress, error: null, done, isSourceValid: importState.isSourceValid, numObj})
+        }
         try {
-            let paginationResp = {}
-            let after = ""
-            let prepend = `${destRef.current.value}`;
-            let importBranchResp
-            let sum = importState.numObj
-            const commitMsg = commitMsgRef.current.value;
-            const sourceRefVal = sourceRef.current.value;
-            const rangeArr = []
-
-            do {
-                const response = await ranges.createRange(repo.id, sourceRefVal, after, prepend, paginationResp.continuation_token)
-                rangeArr.push(response.range)
-                paginationResp = response.pagination
-                after = paginationResp.last_key
-                sum += response.range.count
-                setImportState({inProgress: true, error: null, done: false, isSourceValid: importState.isSourceValid, numObj: sum})
-            } while (paginationResp.has_more);
-            const metarange = await metaRanges.createMetaRange(repo.id, rangeArr)
-
-            try {
-                importBranchResp = await branches.get(repo.id, importBranch)
-            } catch (error) {
-                if (error instanceof NotFoundError) {
-                    // Find root commit for repository
-                    let hasMore = true;
-                    let nextOffset = "";
-                    let baseCommit = reference.id;
-                    do {
-                        let response = await commits.log(repo.id, reference.id, nextOffset, 1000);
-                        hasMore = response.pagination.has_more;
-                        nextOffset = response.pagination.next_offset;
-                        baseCommit = response.results.at(-1);
-                    } while (hasMore)
-                    await branches.create(repo.id, importBranch, baseCommit.id)
-                    importBranchResp = await branches.get(repo.id, importBranch)
-                } else {
-                    setImportState({...initialState, error})
-                    throw error
-                }
-            }
-
-            await commits.commit(repo.id, importBranchResp.id, commitMsg, {}, metarange.id)
-            setImportState({inProgress: false, error: null, done: true, numObj: sum, isSourceValid: importState.isSourceValid})
-            onDone()
+            await runImport(updateStateFromImport,
+                `${destRef.current.value}`,
+                commitMsgRef.current.value,
+                sourceRef.current.value,
+                importBranch,
+                repo.id,
+                reference.id
+            );
+            onDone();
```

Comment:
(Not introduced in this PR but possibly relevant...)

Is `onDone` a synchronous call, or should this have been `await onDone()`?  In particular, if it's asynchronous, then IIRC both this _and the previous version_ lose exceptions that may occur while it runs.

Reply:
`onDone` is a method that's passed from outside.. Don't think we should enforce a wait on it.
Nonetheless, the passed `onDone` function is just a state update function (toggles a boolean)